### ISSUE TITLE
Post processing loop when “keep original files” checked

### ISSUE THREAD

**Issue by barbequesauce** [DobyTang/LazyLibrarian#1217]
To help with identifying and fixing issues, please include as much information as possible, including:
### Relevant debug log: coming....

Wanted to get this out there while I chase a log. Since I turned on “keep original files”, it seems to get stuck in a pp loop, repostprocessing files over and over...

**Comment by barbequesauce** [comment]
[debug.zip](https://github.com/DobyTang/LazyLibrarian/files/1714973/debug.zip)

**Comment by barbequesauce** [comment]
That log should have Change Agent from Daniel Suarez, Ringworld from Larry Niven, and The Atlantis Gene from A. G. Riddle repeating.

**Comment by philborman** [comment]
Log looks ok, only processes each book once. Because the original copies of the books are in the download directory they get checked every time for each _other_ book you look for. The postprocessor checks every folder in downloads for each book it's trying to match. Might be able to keep a list of ones we've already processed so we can skip them.

**Comment by barbequesauce** [comment]
Maybe I’m misunderstanding then - I’m definitely getting Pushover alerts for each book every time it scans the folder, even if they’ve been pp’d already.

**Comment by philborman** [comment]
The log only shows one pushover per book?

**Comment by barbequesauce** [comment]
Try this one... you should see two notifications each for Atlantis Gene and Change Agent. There's also notfications for Ringworld, The Girl who Kicked the Hornets Nest, and Mona Lisa Overdrive - I'm running out the door to catch a flight so I killed it before they re-sent...

[debug.zip](https://github.com/DobyTang/LazyLibrarian/files/1716662/debug.zip)

**Comment by philborman** [comment]
Ok, think I have found the bug, and speeded up postprocessing a bit as a handy bonus :-)
The problem was when a torrent renamed a file when keep-originals was set. We ended up processing the original and the copy, so you got two results for each renamed book. Just need to test it a bit more before releasing an update.

**Comment by barbequesauce** [comment]
Beautiful!

**Comment by philborman** [comment]
Just pushed the new version, hopefully fixed, if not please re-open the issue

**Comment by barbequesauce** [comment]
Sadly no... still happening. Attached debug log should show three pp’s for each of five or so books...
[debug.zip](https://github.com/DobyTang/LazyLibrarian/files/1732269/debug.zip)

**Comment by philborman** [comment]
That's frustrating, thought I'd found it.
Looking through the log though there are only 7 pushover messages, 1 each for 3 books (normal) and 2 each for 2 other books (not normal).  Needs more thought, not as simple as it first looked.
We're missing a couple of useful lines in the debug log 

Found %s file%s marked "Snatched"  
and
Checking %s file%s in %s

which would give a bit more info on what's in the rest of the log. It looks as though some of the books could be in "Snatched" more than once, the line "book x was sent to Transmission" appears more than once for some books. Shouldn't happen. Either we're looking at  more than one postprocessor output, or somehow we've snatched it more than once?

**Comment by barbequesauce** [comment]
Not sure. I only have one provider (MaM via Jackett) enabled, so how would it have multiple snatches?

Also - there are five books in question, and I received each one four times - starting with Mona Lisa Overdrive at 12:14 and about eight minutes between each repeat. That said, I saved the debug log before I killed it, so I’m not sure how many I actually captured in the debug log. I’m happy to send the complete “raw” log if it helps, via email I guess?

**Comment by philborman** [comment]
Not sure if there might be multiple copies marked "wanted" in the database, can't see how we can get that message (book sent to transmission) multiple times in the logs otherwise. What's in your history list, postprocessor should only look for the ones marked "Snatched"

**Comment by barbequesauce** [comment]
You’re correct - there are multiple entries for each of these, at least three for each. The lines seem to be the same except for the time stamps... see attached as an example. Should I clear Snatched to fix it?

![5fb804d4-a891-42c5-89e9-7f1397441c94](https://user-images.githubusercontent.com/25070944/36329145-f2e90580-1332-11e8-9f6d-35bcc2e09d30.jpeg)

**Comment by philborman** [comment]
Yes, and then mark any that genuinely failed as "wanted" again so you retry them. I will need to look into why we snatched them more than once. They are all 3 hours apart, so looks like a new snatch each time the search tasks run. Normally the first one (02:21:50) would succeed and change to "processed", or fail and get markd "failed" (blacklisted). Bug somewhere. I can add code to stop the multiple snatches, but also need to figure out why it happened.

There is also a potential issue with the postprocessor frequency. As you are grabbing quite a few books the postprocessor is taking several minutes to run, and we default to a short interval of 10 minutes on the assumption that it doesn't take long to download a book. One scenario could be that one postprocessor is running, and 10 minutes later another one starts. Needs to be a flag to tell it not to.

**Comment by barbequesauce** [comment]
The delay is explainable - MaM limits new accounts to have only x number of downloads activ over y days. When you first join it’s limited to 20 for 3 day’s, and that accelerates over time.....

Not sure how you would implement a fix for this, just trying to illustrate the problem a bit better...

**Comment by philborman** [comment]
The MaM limit isn't a problem, it's just that if you are postprocessing a lot of books, another postprocessor might start before the first one has finished. Easy enough to block it.

I think the multple entries might be a difference between jackett and "normal" providers. We use the URL as an identifier in the history table, and I think jackett maybe gives you a unique ID each search, even if you search for the same book. We can confirm this easily enough by looking at the history table. 

If you haven't already cleared snatched, can you try using the api command getHistory, that's the easiest way to get the info, otherwise we can use sqlite3 from the commandline. I think you will find for book titles with multiple entries the NZBurl is different for each entry.

**Comment by barbequesauce** [comment]
Sorry, not familiar with how to perform api commands?

**Comment by philborman** [comment]
You might not have api enabled by default, there is a tick box in config and a random access key.

You access from a browser just like getting the lazylibrarian Web pages, but with the api key and a command like this... 

http://192.168.1.2/lazylibrarian/api?apikey=e136f5553b7f83a1b579a20a60&cmd=getHistory

Results look like this... 

[{"Status": "Processed", "NZBprov": "https://drunkenslug.com", "AuxInfo": "2018-03-01", "NZBmode": "nzb", "DownloadID": "SABnzbd_nzo_b_QpUg", "Source": "SABNZBD", "NZBsize": "56.77", "NZBtitle": "Super.Food.Ideas.TruePDF-March.2018", "NZBurl": "https://drunkenslug.com/getnzb/139939cbfa3e3542dac6286bb46a0930035cf03d.nzb&i=11000&r=", "BookID": "Super Food Ideas", "NZBdate": "2018-02-09 21:03:57"}, {"Status": "Processed", "NZBprov": "https://drunkenslug.com", "AuxInfo": "2018-03-01", "NZBmode": "nzb", "DownloadID": "SABnzbd_nzo_AJaAsk", "Source": "SABNZBD", "NZBsize": "65.79", "NZBtitle": "Family.Circle.TruePDF-March.2018", "NZBurl": "https://drunkenslug.com/getnzb/7c865f03d74cc1b9eec125f327a6123ccb214584.nzb&i=11000&r=", "BookID": "family circle", "NZBdate": "2018-02-14 11:00:19"}, {"Status": "Processed", "NZBprov": "https://drunkenslug.com", "AuxInfo": "2018-03-01", "NZBmode": "nzb", "DownloadID": "SABnzbd_nzo_D055ZK", "Source": "SABNZBD", "NZBsize": "79.07", "NZBtitle": "House.Beautiful.USA.TruePDF-March.2018", "NZBurl": "https://drunkenslug.com/getnzb/4d2b65a9f87b61f836217bc0b6d196ef.nzb&i=11000&r=", "BookID": "House Beautiful", "NZBdate": "2018-02-14 11:00:25"}, {"Status": "Failed", "NZBprov": "https://drunkenslug.com", "AuxInfo": "AudioBook", "NZBmode": "nzb", "DownloadID": "SABnzbd_nzo_3BoiGC", "Source": "SABNZBD", "NZBsize": "882.95", "NZBtitle": "Stieg Larsson - The Girl with the Dragon Tattoo LL.(2429135)", "NZBurl": "https://drunkenslug.com/getnzb/c6a87b7f9b4f5adcdbd2780f8c6561f9.nzb&i=11000&r=", "BookID": "2429135", "NZBdate": "2018-02-14 22:10:23"}]

The thing to check is that I'm guessing that for multiple entries for the same book,  the nzb url is different for each entry?

**Comment by barbequesauce** [comment]
OK - full getHistory output [here](https://gist.github.com/barbequesauce/eed7f08c885293e87ad5dc4f7df92c1f)... I think these three lines are examples of what you're looking for?

TL;DR: yes the Jackett URLs are unique - more specifically, the path argument in the URL.


```
{"Status": "Processed", "DownloadID": null, "BookID": null, "Source": null, "NZBmode": null, "AuxInfo": null, "NZBdate": "2018-02-16 18:05:12", "NZBtitle": null, "NZBsize": null, "NZBurl": "http://piratepi.local:9090/jackett/dl/myanonamouse/?jackett_apikey=YouDontNeedThis&path=VHE3ZWJxOWF2RmphaUl3RmE1aEx6MWZrbkdCYUhFa1hxNUllenlrL0oxdHBCVDBBR09MRVlzSjZ5aWo4dXBMZmpST0tnRFM0L3B6UlJKeXlVZlgyaWV4T0wyNWVvdVFFSEFDVHArZnVqZXdkOW5hamt4MEI0OGQ3YmFwT1NTekVCSm9zek94Tis3eHEyWkxrVE5VUm4xa0tnSTY4THJkWHF0U3JlQ0NKYzFNSzl2YVpFOEFOR2xOSVVPWDBTSHNhSEppdUtnV1BBS2dMMEhvQU1pY1piMlI2dWdMQmxXYkQ0UDc0Q2FOWi9OVEhZM3B0cXdBSWs2T0QvRllDTktkM2F6U2t2N2lnUGNxYmxlbXVodVcwTmFycHlnbGJla1lVU2xvRU1ienU1ODRNZ01pWW9RR0VNM3JpZW9xZ2t1RnBidElXL01FTXNNUDIwaTd3M0lQcjNNVnYxQzBvSHJrQzlGZEtHeG03VXM4PQ2&file=Ringworld+by+Larry+Niven", "NZBprov": null}
{"Status": "Processed", "DownloadID": null, "BookID": null, "Source": null, "NZBmode": null, "AuxInfo": null, "NZBdate": "2018-02-16 18:10:47", "NZBtitle": null, "NZBsize": null, "NZBurl": "http://piratepi.local:9090/jackett/dl/myanonamouse/?jackett_apikey=YouDontNeedThis&path=eFJndlVxaGVlVTZEbHlCSWc3ZjV4SmpDOTRKVWM5eXFkdFcvY2t6OCtlZjZacHNXVktjcHJjY2ZkZXRuTDFZMVVOSGxUc2gwcENkaDFwczVhQy8wamIwSDRRNHg3N3hGYzM5OWQ5MnpXUTNVaFlyZEFTUHZNV3dqanhwa1dLdGo3YmpKMXhWMHNzR2cwUEhWejRGYnZ3MUN2UUw1THN3ZStpVTdKeGxXSFNhNzc5cEU4WFY5RlpEaENpRkxjZGpzdVR6Z1JsbzErdmxoWDB1MlRoWWlQTXdJTGRmVW5vQUY4VWZRVjdodVB2Q3hndWc1TE1MQ3Jaa01nUmNtMVdpS25NT2trR3VoNkVaQkV4NTl5TWhzTlBQWjlwT2ZkT0FWZlNqUWd6OVovVWNubTNFaFRqU29ZRWVuVUUzTERYVW9mQitkMExISi84UVYzMjgzbW54cVNyTlNLYnpDNXFGVm54Zm01WGswaFJvPQ2&file=Ringworld+by+Larry+Niven", "NZBprov": null}, 
 {"Status": "Processed", "DownloadID": null, "BookID": null, "Source": null, "NZBmode": null, "AuxInfo": null, "NZBdate": "2018-02-16 18:16:27", "NZBtitle": null, "NZBsize": null, "NZBurl": "http://piratepi.local:9090/jackett/dl/myanonamouse/?jackett_apikey=YouDontNeedThis&path=L0NkYlJMSXJ3ZjV4czdRTDYzandORElMMlBOVFJ4ZTgzTG9ieFNCWGpIbXpZVU9FY3l2TENtWTJCNlA4ZDBEQ0o2ZWxRWkx4SkZ3Vy9jK2FyZFhJeW5PdWpxL3p6Z3VPRUFtQnc5SEg2bDN6NzRnNmI1c0xFajlxSklhck5rZGFuYnUrb1YyZEVRZk12TUdVM0k0QlNHT2dzNW9JK0t6bjRrS01yT1BDNHJudEN4ekRVRkpqd041YlQwSXpzSHNsL0xESktsTW03eTFwL1RzTjdDMG5GK0NQdUpRV2xBbUhRNEd1bi85VDVVWXA0bjRFN201WEVaTmdzejR1dC9Bb0NlQWZJaHZ4VlJHQWdHMm5jQ3k3VkRpTzVsQkNTNlNtNjAxdyt3elpsNDRKaFZDWGRqRHNVMDFURnJnakM4WE43RGFXNEtjc2JmRWZWYnpsRXgxNDk0MUoyWUplU3VGQUdYaVhKZjZTU3ZZPQ2&file=Ringworld+by+Larry+Niven", "NZBprov": null},
```

**Comment by philborman** [comment]
Yes that's excellent thanks. Just what I was expecting. 
I can code around the problem fairly easily now that's confirmed.
Jackett (and maybe other proxies, Cardigann, NZBHydra) gives you a different url each time for the same search. We currently use the url as an identifier, obviously won't work if the identifier changes.
I can add some code to check for a combination of bookid and type instead.  Thanks for the help.

**Comment by barbequesauce** [comment]
Thank you for jumping on it! Not an easy problem to dig into.

**Comment by philborman** [comment]
Just pushed a new version with two checks,  so we should catch using the url as before, but also before searching we check using the bookid and type (ebook / audiobook) and don't search if it's already marked as snatched.

**Comment by barbequesauce** [comment]
Looks good, thanks!
